csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,116 @@ void insertReshardings(Fusion* fusion) {
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      IterDomain* shard_deleted_id = shard_deletions[0];
+      int sharding_axis =
+          static_cast<int>(input->domain()->rootPosOf(shard_deleted_id));
+      if (isOutermostAllocatedId(input, shard_deleted_id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
+      TensorView* output_permute = set(input_permute);
+      TensorView* new_output = permute(output_permute, {{0, sharding_axis}});
+      ir_utils::replaceValInAllExprInputsAndFusionOutputs(output, new_output);
+
+      // Propagate shardings from input and manually apply sharding deletions.
+      shardAllLike(input, {input_permute, output_permute, new_output});
+      output_permute->axis(0)->parallelize(ParallelType::Serial);
+      new_output->axis(sharding_axis)->parallelize(ParallelType::Serial);
+      output_permute->setDeviceMesh(output->getDeviceMesh());
+      new_output->setDeviceMesh(output->getDeviceMesh());
+    }
+    // For scatter operations i.e. ID goes from unsharded to sharded
+    // Update input to push the scattered axis to the front -> collective ->
+    // permute the sharded axis to the proper location.
+    // Example: [i0 DIDx(i1) i2] -> [i0 r1 DIDx(i2)]
+    // Rewritten to: [i0 DIDx(i1) i2] -> [i2 i0 DIDx(i1)] ->
+    //                    [DIDx(i2) i0 r1] -> [i0 DIDx(i2)]
+    // Note that reduction axis shifts from axis=1 to axis=2.
+    else if (!shard_additions.empty()) {
+      auto shard_added_id = shard_additions[0];
+      int sharding_axis =
+          static_cast<int>(output->domain()->rootPosOf(shard_added_id));
+      if (isOutermostAllocatedId(output, shard_added_id)) {
+        continue;
+      }
+
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
+      TensorView* output_permute = nullptr;
+      // Calculate the number of reduction axes before the sharding axis.
+      // After permuting the sharding axis to the front, the reduction axis
+      // will be offset by this amount.
+      auto reduction_axis = output->getReductionAxis();
+      int num_reduction_axes_before_sharding_axis =
+          (reduction_axis.has_value() &&
+           sharding_axis > static_cast<int>(reduction_axis.value()))
+          ? 1
+          : 0;
+      if (expr->isA<ReductionOp>()) {
+        int reduction_axis_after_permute =
+            static_cast<int>(reduction_axis.value()) +
+            num_reduction_axes_before_sharding_axis;
+        auto red_expr = dynamic_cast<ReductionOp*>(expr);
+        output_permute = reductionOp(
+            red_expr->getReductionOpType(),
+            {reduction_axis_after_permute},
+            red_expr->init(),
+            input_permute);
+      } else {
+        // Note this is a no-op and is moving a device parallel axis back
```

Comment:
I am not sure to understand this comment.
It is not a no-op since it embedds a scatter, right ?

Reply:
good catch, the permute after this is the view change